src/simsopt/geo/permanent_magnet_grid.py

```diff
@@ -0,0 +1,549 @@
+
+        # Extra work below so that the stitching with the symmetries is done in
+        # such a way that the reflected cells are still dx and dy away from
+        # the old cells.
+        #### Note that Cartesian cells can only do nfp = 2, 4, 6, ... 
+        #### and correctly be rotated to have the right symmetries
+        if (self.plasma_boundary.nfp % 2) == 0:
+            X = np.linspace(self.dx / 2.0, (x_max - x_min) + self.dx / 2.0, Nx, endpoint=True)
+            Y = np.linspace(self.dy / 2.0, (y_max - y_min) + self.dy / 2.0, Ny, endpoint=True)
+        else:
+            X = np.linspace(x_min, x_max, Nx, endpoint=True)
+            Y = np.linspace(y_min, y_max, Ny, endpoint=True)
+        Z = np.linspace(-z_max, z_max, Nz, endpoint=True)
+        print(Nx, Ny, Nz, X, Y, Z, self.dx, self.dy, self.dz)
+
+        # Make 3D mesh
+        X, Y, Z = np.meshgrid(X, Y, Z, indexing='ij')
+        self.xyz_uniform = np.transpose(np.array([X, Y, Z]), [1, 2, 3, 0]).reshape(Nx * Ny * Nz, 3)
+
+        # Extra work for nfp = 4 to chop off half of the originally nfp = 2 uniform grid
+        if self.plasma_boundary.nfp == 4:
+            inds = []
+            for i in range(Nx):
+                for j in range(Ny):
+                    for k in range(Nz):
+                        if X[i, j, k] < Y[i, j, k]:
+                            inds.append(int(i * Ny * Nz + j * Nz + k))
+            good_inds = np.setdiff1d(np.arange(Nx * Ny * Nz), inds)
+            self.xyz_uniform = self.xyz_uniform[good_inds, :]
+        contig = np.ascontiguousarray
+        pointsToVTK(
+            'uniform_grid',
+            contig(self.xyz_uniform[:, 0]),
+            contig(self.xyz_uniform[:, 1]),
+            contig(self.xyz_uniform[:, 2])
+        )
+
+    def geo_setup_from_famus(self, famus_filename, m_maxima=None):
+        """
+        Function to initialize a SIMSOPT PermanentMagnetGrid from a 
+        pre-existing FAMUS file defining a grid of magnets. For
+        example, this function is used for the
+        half-Tesla NCSX configuration (c09r000) and MUSE grids.
+        If this flag is specified, rz_inner_surface,
+        rz_outer_surface, dr
+        are all ignored.
+
+
+        Args
+        ----------
+        famus_filename : string
+            Filename of a FAMUS grid file (a pre-made dipole grid).
+        m_maxima: float or 1D numpy array, shape (Ndipoles)
+            Optional array of maximal dipole magnitudes for the permanent
+            magnets. If not provided, defaults to the common magnets
+            used in the MUSE design, with strengths ~ 1 Tesla.
+
+        Returns
+        -------
+        None.
+
+        """
+        if famus_filename is not None:
+            warnings.warn(
+                'famus_filename variable is set, so a pre-defined grid will be used. '
+                ' so that the following parameters are ignored: '
+                'rz_inner_surface, rz_outer_surface, dr.'
+            )
```

Comment:
If famus_file has rz_inner_surface, rz_outer_surface, etc, why can't we use them?